File: src-tauri/src/app/modules/core/remove.rs
use crate::app::modules::{core::ops::manager::ModManager, manifest::models::Provider};

pub struct Remove;

impl Remove {
    pub async fn run(slug: String, provider: Option<Provider>) -> Result<Option<String>, String> {
        let mut manager = ModManager::new();
        manager.load().await;
        manager.with_default_providers(); // TODO: Check if required

        let provider = provider.unwrap_or(manager.manifest_service.manifest.default_provider.clone());
        if !manager.manifest_service.manifest.remove_mod_entry(&provider, &slug) {
            return Ok(Some(format!("Mod '{}' not found in manifest", slug)));
        }

        manager.lock_service.lock.mods.remove(&slug);
        manager
            .save_all()
            .map_err(|e| format!("Failed to save state: {}", e))?;

        Ok(None)
    }
}

File: packages/core/src/operations/remove.ts
import type {ModManager} from "./mod-manager.js";
import type {Provider, ResourceKind} from "../models/manifest.js";
import {removeModEntry} from "../models/manifest.js";

export class Remove {
  static async run(
    manager: ModManager,
    slug: string,
    provider?: Provider,
    kind: ResourceKind = "mod",
  ): Promise<string | undefined> {
    const resolvedProvider = provider ?? manager.manifestService.manifest.default_provider;
    const label = kind === "datapack" ? "Datapack" : "Mod";

    if (!removeModEntry(manager.manifestService.manifest, resolvedProvider, slug, kind)) {
      return `${label} '${slug}' not found in manifest`;
    }

    const key = `${resolvedProvider}:${slug}`;
    const lockMap = kind === "datapack" ? manager.lockService.lock.datapacks : manager.lockService.lock.mods;
    lockMap.delete(key);
    manager.saveAll();

    return undefined; // success, no warning
  }
}
